src/PrestaShopBundle/Form/Admin/Sell/Product/VirtualProductFileType.php

```diff
@@ -143,5 +153,28 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'empty_data' => '',
             ])
         ;
+
+        // Preset the input attributes correctly depending on the data
+        $builder->addEventListener(FormEvents::PRE_SET_DATA, function (FormEvent $event) {
+            $data = $event->getData();
+
+            if (empty($data['virtual_product_file_id'])) {
+                return;
+            }
+            $virtualProductFileId = (int) $data['virtual_product_file_id'];
+            $virtualProductFileDownloadUrl = $this->router->generate('admin_products_v2_download_virtual_product_file', [
+                'virtualProductFileId' => $virtualProductFileId,
+            ]);
+
+            $form = $event->getForm();
+            $fileField = $form->get('file');
+            $fileOptions = $fileField->getConfig()->getOptions();
+            $fileOptions['download_url'] = $virtualProductFileDownloadUrl;
+
+            // Replace existing field with new one with adapted options
+            $cloner = new FormCloner();
```

Comment:
why do we need this? looks like some workaround :thinking:

Reply:
Hi, it's not at all a workaround I wish there was a simpler way but to generate the URL we need the `virtual_product_file_id` from the data, problem is when the form is built we don't have the data accessible in the config So we need to use this event to correctly modify the form when the data is set

So I listen to the PRE_SET_DATA event, but when this event is triggered the Form is already built, and there is no simple way to modify an option, except rebuild the field itself and add it again inside the built form That's where the FormCloner is handy, it's a tool that was developed to allow cloning a form field and be sure that nothing is dropped while copying it (options, listeners, ...)